weekly-report/main.py

```diff
@@ -1 +1,134 @@
 "A script to generate weekly PDF reporting"
+from datetime import timedelta, date
+from os import environ as ENV
+import os
+from io import BytesIO
+import logging
+from base64 import b64encode
+from xhtml2pdf import pisa
+import altair as alt
+from boto3 import client
+from botocore.exceptions import NoCredentialsError
+
+from visuals import get_state_risk_map, get_magnitude_map
+from visuals import get_significance_bar
+
+CURRENT_DATE = date.today()
+DESTINATION_DIR = '/tmp/data'
+
+
+def get_s3_client() -> client:
+    """Returns input s3 client"""
+    try:
+        s3_client = client('s3',
+                           aws_access_key_id=ENV.get('ACCESS_KEY'),
+                           aws_secret_access_key=ENV.get('SECRET_ACCESS_KEY'))
+        return s3_client
+    except NoCredentialsError:
+        logging.error("Error, no AWS credentials found")
+        return None
+
+
+def get_file_keys_from_bucket(s3, bucket_name: str) -> list[str]:
+    """Gets all file Key values from a bucket"""
+    bucket = s3.list_objects(Bucket=bucket_name)
+    return [file["Key"] for file in bucket["Contents"]]
+
+
+def download_shapefiles(bucket_name: str, s3: client, folder_path: str) -> None:
+    """Downloads shapefiles for US state mapping"""
+    file_keys = get_file_keys_from_bucket(s3, bucket_name)
+    for file_key in file_keys:
+        file_path = os.path.join(folder_path, file_key)
+        os.makedirs(os.path.dirname(file_path), exist_ok=True)
+        s3.download_file(bucket_name, file_key, file_path)
+
+
+def convert_altair_chart_to_html_embed(chart: alt.Chart) -> str:
+    """Converts an Altair chart to a string representation."""
+
+    with BytesIO() as bs:
+        chart.save(bs, format="png")
+        bs.seek(0)
+
+        data = b64encode(bs.read()).decode("utf-8")
+
+    return f"data:image/png;base64,{data}"
+
+
+def convert_html_to_pdf(source_html) -> BytesIO:
+    """Outputs HTML to a target file."""
+    pdf_output = BytesIO()
+    pisa_status = pisa.CreatePDF(
+        source_html,
+        dest=pdf_output)
+
+    if pisa_status.err:
+        raise Exception("error converting html to PDF")
+    pdf_output.seek(0)
+
+    return pdf_output
+
+
+def get_prefix(current_date: date) -> str:
+    """creates object prefix for s3 bucket"""
+    weekday = current_date.weekday()
+    if weekday == 0:
+        return f"wc-{current_date.strftime("%d-%m-%Y")}/"
+
+    monday = current_date - timedelta(weekday)
+    return f"wc-{monday.strftime("%d-%m-%Y")}/"
+
+
+def upload_historical_readings(s3: client, bucket_name: str, prefix: str,
+                               filename: str, file_data: BytesIO) -> None:
+    """uploads the historical readings in memory to the s3 bucket"""
+    object_key = os.path.join(prefix, filename)
+    s3.upload_fileobj(file_data, bucket_name, object_key)
+    logging.info(f"csv file uploaded: {object_key}")
+
+
+def handler(event=None, context=None):
```

Comment:
Good work breaking down the work for this program into smaller functions, it makes it much easier to read the purpose of this program.